### PR TITLE
Admin router Marathon 1.5 container networking support

### PR DESCRIPTION
**Pull request by ti-mo**
## High-level description

With the transition to Marathon 1.5 in DC/OS 1.10.x, this code was not updated to reflect the changes in Marathon task/app API (no more `container.docker.network`, `networks` being at the root of the app definition, etc.).

The IP-per-task concept is now also marked as deprecated in the Marathon documentation, so that's been removed except for one test (`test_ip_per_task_app_with_unspecified_ip_address_DCOS_OSS_1366`) that I didn't want to touch. 

`is_ip_per_task` and `is_user_network` were condensed into one simple check, `is_container_network()`. `container.type` does no longer need to be `DOCKER`, since I assume the container networking concept will carry over to other containerizers, too. (probably the whole reason for this change)

`test_ip_per_task_app_without_user_networking` was removed, since I don't think this is a possible scenario anymore? (correct me if I'm wrong)

## Corresponding DC/OS tickets (obligatory)

This work is a spontaneous bug fix introducing a minor change, only concerning the admin UI.

## Related tickets (optional)

Other tickets related to this change:

  - Not a JIRA ticket, but: https://github.com/dcos/dcos/pull/1628

## Checklist for all PRs

  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]

### CODE REVIEW

**Comment by ti-mo** [comment]
@jgehrcke @lingmann @vespian

File: packages/adminrouter/extra/src/lib/cache.lua [hunk 1]
```diff
@@ -197,22 +194,17 @@ local function fetch_and_store_marathon_apps(auth_token)
...
-         ports = {}
-         if is_user_network(app) then
+
+       if is_container_network(app) then
             -- override with ports from the container's portMappings
-            local port_mappings = app["container"]["docker"]["portMappings"] or app["portDefinitions"] or {}
-            local port_attr = app["container"]["docker"]["portMappings"] and "containerPort" or "port"
+            ports = {}
+
+            local port_mappings = app["container"]["portMappings"] or app["portDefinitions"] or {}
```
Comment (jdef) [inline packages/adminrouter/extra/src/lib/cache.lua]: there won't be any `portDefinitions` if the app is running on a container network

File: packages/adminrouter/extra/src/lib/cache.lua [hunk 2]
```diff
@@ -197,22 +194,17 @@ local function fetch_and_store_marathon_apps(auth_token)
...
+            local port_attr = app["container"]["portMappings"] and "containerPort" or "port"
```
Comment (jdef) [inline packages/adminrouter/extra/src/lib/cache.lua]: `port` is not an attribute of `ContainerPortMapping`: https://github.com/mesosphere/marathon/blob/3ff7f927693fd208b177e53fcdc716e1b647f344/docs/docs/rest-api/public/api/v2/types/appContainer.raml#L43

File: packages/adminrouter/extra/src/test-harness/tests/test_cache.py [hunk 3]
```diff
@@ -899,13 +899,16 @@ def test_if_backend_requests_have_useragent_set_correctly(
...
 class TestCacheMarathon:
 
-    def test_ip_per_task_app_with_user_networking_and_portmappings(
+    def test_app_with_container_networking_and_portmappings(
             self, nginx_class, mocker, valid_user_header):
+
         app = self._scheduler_alwaysthere_app()
-        app['ipAddress'] = {'networkName': 'samplenet'}
+
+        app['networks'][0]['mode'] = 'container'
```
Comment (jdef) [inline packages/adminrouter/extra/src/test-harness/tests/test_cache.py]: i'm confused by the `[0]` here. in other places I see `[1]` to access the first element. probably my lack of lua expertise

Comment (ti-mo) [reply packages/adminrouter/extra/src/test-harness/tests/test_cache.py]: Correct, lua is 1-indexed.

**Comment by vespian** [comment]
@ti-mo  Is this ready for the next pass?

**Comment by ti-mo** [comment]
@vespian Yes, I've reduced the change set to the minimum following @jdef's remarks. There were some unused/unnecessary code paths. Can't get the test suite to run properly, though. A large part of the test suite on master is failing here, is that normal?

**Comment by ti-mo** [comment]
@vespian I've pushed another commit that removes the test cases with `portDefinition`s (not used for container networks) and fixes the `portMappings` test, at least I got those to pass. :)

File: packages/adminrouter/extra/src/lib/cache.lua [hunk 4]
```diff
@@ -95,13 +95,9 @@ local function request(url, accept_404_reply, auth_token)
...
 
-local function is_ip_per_task(app)
-    return app["ipAddress"] ~= nil and app["ipAddress"] ~= cjson_safe.null
-end
-
-local function is_user_network(app)
+local function is_container_network(app)
     local container = app["container"]
-    return container and container["type"] == "DOCKER" and container["docker"]["network"] == "USER"
+    return container and app["networks"][1]["mode"] == "container"
```
Comment (vespian) [inline packages/adminrouter/extra/src/lib/cache.lua]: `networks` key may be not present[1], the specification does not mention that there is going to be at least one network entry in the `app["networks"][1]` list[1][2]


[1] https://github.com/mesosphere/marathon/blob/5d2b7970c147e11c8181971f381b179e773d3671/docs/docs/rest-api/public/api/v2/types/app.raml#L239-L241
[2] https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md#array-type

File: packages/adminrouter/extra/src/lib/cache.lua [hunk 5]
```diff
@@ -179,8 +175,9 @@ local function fetch_and_store_marathon_apps(auth_token)
           )
 
        local host_or_ip = task["host"] --take host  by default
-       if is_ip_per_task(app) then
-          ngx.log(ngx.NOTICE, "app '" .. appId .. "' is using ip-per-task")
+
+       if is_container_network(app) then
+          ngx.log(ngx.NOTICE, "app '" .. appId .. "' is in container network")
```
Comment (vespian) [inline packages/adminrouter/extra/src/lib/cache.lua]: `s/is in container/is in a container/`

File: packages/adminrouter/extra/src/test-harness/tests/test_cache.py [hunk 6]
```diff
@@ -924,12 +927,16 @@ def test_ip_per_task_app_with_user_networking_and_portmappings(
             req_data = resp.json()
             assert req_data['endpoint_id'] == 'http://127.0.0.2:80'
 
-    def test_ip_per_task_app_with_user_networking_and_portdefinitions(
+    def test_app_with_container_networking_and_portdefinitions(
```
Comment (vespian) [inline packages/adminrouter/extra/src/test-harness/tests/test_cache.py]: Can we also test the edge cases that I mentioned in my first comment?
* networks key may be not present
* the specification does not mention that there is going to be at least one network entry in the app["networks"][1]

**Comment by kensipe** [comment]
any status update?

**Comment by ti-mo** [comment]
@kensipe @vespian Sorry about the slow turnaround, I believe I addressed the feedback.

**Comment by jgehrcke** [comment]
@mesosphere-mergebot label Ready For Review

**Comment by jgehrcke** [comment]
I would like to see a concise problem statement somewhere, ideally in a JIRA ticket. I have read this:

> With the transition to Marathon 1.5 in DC/OS 1.10.x, this code was not updated to reflect the changes in Marathon task/app API (no more container.docker.network, networks being at the root of the app definition, etc.).

But it does not really explain to me which problem we now have in DC/OS. This sounds serious, but what exactly is broken?

**Comment by ti-mo** [comment]
@jgehrcke @vespian @kensipe Any interest in merging this?

**Comment by jdef** [comment]
It looks like we still need/want a dcos integration test for this as per https://github.com/dcos/dcos/pull/2497#pullrequestreview-100145643

Is that something that you need help with @ti-mo ?

**Comment by jgehrcke** [comment]
@mattj-io @pleia2 We might need help here.

**Comment by mattj-io** [comment]
Sure, what do you need @jgehrcke ?

**Comment by ti-mo** [comment]
@jdef @jgehrcke No time budget to work on this at the moment. I am not familiar with the test suite; given clear instructions on how to run and test this locally, I might be able to get something going. Last time I tried, it was not very friendly, the documentation was not up-to-date, and most (unit) tests were broken on master.

I think it would be more efficient if a core contributor could chip in and write a test for this, I expect this is trivial for someone familiar with the codebase.

**Comment by orsenthil** [comment]
If we want this change, this should be taken by a DC/OS developer - @vespian

**Comment by orsenthil** [comment]
@mesosphere-mergebot label work in progress.
